src/arch_ppc.py

```diff
@@ -0,0 +1,862 @@
+from typing import (
+    ClassVar,
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+)
+from .error import DecompFailure
+from .options import Target
+from .parse_instruction import (
+    Argument,
+    AsmAddressMode,
+    AsmGlobalSymbol,
+    AsmLiteral,
+    Instruction,
+    InstructionMeta,
+    JumpTarget,
+    Macro,
+    Register,
+)
+from .asm_pattern import (
+    AsmMatch,
+    AsmMatcher,
+    AsmPattern,
+    Replacement,
+    SimpleAsmPattern,
+    make_pattern,
+)
+from .translate import (
+    Abi,
+    AbiArgSlot,
+    Arch,
+    BinaryOp,
+    Cast,
+    CmpInstrMap,
+    CommentStmt,
+    ErrorExpr,
+    Expression,
+    ImplicitInstrMap,
+    InstrMap,
+    InstrSet,
+    PpcCmpInstrMap,
+    PairInstrMap,
+    SecondF64Half,
+    StmtInstrMap,
+    StoreInstrMap,
+    UnaryOp,
+    as_f32,
+    as_f64,
+    as_int64,
+    as_intish,
+    as_intptr,
+    as_ptr,
+    as_s32,
+    as_type,
+    as_u32,
+    fn_op,
+    fold_divmod,
+    fold_mul_chains,
+    handle_add,
+    handle_add_double,
+    handle_add_float,
+    handle_addi,
+    handle_addis,
+    handle_convert,
+    handle_load,
+    handle_loadx,
+    handle_or,
+    handle_rlwinm,
+    handle_sra,
+    load_upper,
+    make_store,
+    make_storex,
+    void_fn_op,
+)
+from .types import FunctionSignature, Type
+
+LENGTH_TWO: Set[str] = {
+    "neg",
+    "not",
+}
+
+LENGTH_THREE: Set[str] = {
+    "addi",
+    "ori",
+    "and",
+    "or",
+    "nor",
+    "xor",
+    "andi",
+    "xori",
+    "sllv",
+    "srlv",
+    "srav",
+}
+
+
+class FcmpoCrorPattern(SimpleAsmPattern):
+    pattern = make_pattern(
+        "fcmpo $cr0, $x, $y",
+        "cror 2, N, 2",
+    )
+
+    def replace(self, m: AsmMatch) -> Optional[Replacement]:
+        fcmpo = m.body[0]
+        assert isinstance(fcmpo, Instruction)
+        if m.literals["N"] == 0:
+            return Replacement([m.derived_instr("fcmpo.lte", fcmpo.args)], len(m.body))
+        elif m.literals["N"] == 1:
+            return Replacement([m.derived_instr("fcmpo.gte", fcmpo.args)], len(m.body))
```

Comment:
Add .fictive suffixes -- I was confused by those instructions when I saw them below, and knowing that they were made up would have helped. A doc comment for this class would be good, because I don't understand what this pattern is for.

Reply:
(still missing a doc comment)